### PR TITLE
feat: Add search for votes

### PR DESCRIPTION
**Pull request by j0hnfl0w**
Closes: https://github.com/snapshot-labs/snapshot/issues/3519

- [x] SpaceProposalVotesModal in SpaceProposalPage
- [x] Infinite scroll with useScroll from vueuse
- [x] Search with ETH and ENS addresses

## Changes
Added maxWidth prop to BaseModal
Added autofocus prop to BaseSearch

## How to test
Open any proposal page and click "Search" or "See more"

image
image

## TBD
Its hard to click "See more" at the end of 10 votes on proposal, so I added "Search" on the top of Votes block

### CODE REVIEW

File: src/components/SpaceProposalPage.vue [hunk 1]
```diff
@@ -277,5 +278,17 @@ onMounted(() => {
       :selected-choices="selectedChoices"
       @close="isModalPostVoteOpen = false"
     />
+    <SpaceProposalVotesModal
```
Comment (samuveth) [inline src/components/SpaceProposalPage.vue:281]: This should be inside `SpaceProposalVotesList` and you should use all the sorting login from there instead of repeating is inside the modal again

Comment (j0hnfl0w) [reply src/components/SpaceProposalPage.vue:281]: If we want to do it in SpaceProposalVotesList where we have only 10 votes, then we create there another array of votes for searching.

File: src/components/SpaceProposalVotesModal.vue [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+              v-if="vote.reason !== '' && props.proposal.privacy !== 'shutter'"
+              v-tippy="{
+                content: vote.reason
+              }"
+              class="cursor-default p-0"
+            >
+              <i-ho-annotation class="text-[16px]" />
+            </BaseButtonIcon>
+          </div>
+        </div>
```
Comment (samuveth) [inline src/components/SpaceProposalVotesModal.vue:219]: This is also again repeating code that we already have. We should refactor this into a component `SpaceProposalVotesListItem` and use it both here and in `SpaceProposalVotesList`.

File: src/components/SpaceProposalVotesModal.vue [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+    }
+  }
+);
+</script>
+
+<template>
+  <BaseModal
+    :open="open"
+    max-height="500px"
+    max-width="700px"
```
Comment (samuveth) [inline src/components/SpaceProposalVotesModal.vue:145]: @bonustrack could you comment on design. I'm pretty sure we don't want to change the modal width correct?

Also we want to keep the search and modal aligned like everywhere else I think 
image

File: src/components/SpaceProposalPage.vue (hunk 1 above)
Comment (samuveth) [reply src/components/SpaceProposalPage.vue:281]: Hm yes good point, but we should try to always reuse code where possible.

Comment (samuveth) [reply src/components/SpaceProposalPage.vue:281]: Maybe a composable would make sense

File: src/components/SpaceProposalVotesList.vue [hunk 4]
```diff
@@ -87,14 +87,22 @@ const { downloadVotes, isDownloadingVotes } = useReportDownload();
...
     <template v-if="props.proposal.state === 'closed'" #button>
-      <BaseButtonIcon>
-        <LoadingSpinner v-if="isDownloadingVotes" />
-        <i-ho-download
-          v-else
-          v-tippy="{ content: 'Download as CSV' }"
-          @click="downloadVotes(proposal.id, proposal.space.id)"
-        />
-      </BaseButtonIcon>
+      <div class="flex flex-row items-center">
```
Comment (samuveth) [inline src/components/SpaceProposalVotesList.vue:90]: Using an icon here might be better, or icon with text

File: src/components/SpaceProposalVotesModal.vue (hunk 3 above)
Comment (j0hnfl0w) [reply src/components/SpaceProposalVotesModal.vue:145]: I changed modal maxWidth to preserve Votes block with from Proposal page to make it the same.

File: src/components/SpaceProposalVotesModal.vue [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+              <i-ho-annotation class="text-[16px]" />
+            </BaseButtonIcon>
+          </div>
+        </div>
+        <a
+          v-if="sortedVotes.length < voteCount"
+          class="block rounded-b-none border-t px-4 py-3 text-center md:rounded-b-md"
+          @click="loadMore(loadMoreVotes)"
+        >
+          <LoadingSpinner v-if="loadingMore" />
```
Comment (samuveth) [inline src/components/SpaceProposalVotesModal.vue:225]: The infinite scroll is buggy for me, sometimes when I scroll down I just see "See more" and sometimes I see the spinner.

File: src/components/SpaceProposalPage.vue (hunk 1 above)
Comment (j0hnfl0w) [reply src/components/SpaceProposalPage.vue:281]: When do you start reuse code? With >1 pieces of duplication code or >=3?

File: src/components/SpaceProposalVotesModal.vue (hunk 5 above)
Comment (j0hnfl0w) [reply src/components/SpaceProposalVotesModal.vue:225]: Yep I think at the end there will be only spinner

**Comment by samuveth** [comment]
> Its hard to click "See more" at the end of 10 votes on proposal, so I added "Search" on the top of Votes block

We could reduce the number of votes you see initially to 5-6, that would make the "See more" more visible. I like the idea of adding it in the top but it's not that visible and seems a bit cluttered with the download button.

File: src/components/BaseSearch.vue [hunk 6]
```diff
@@ -54,6 +56,7 @@ function clearInput() {
       type="text"
       autocorrect="off"
       autocapitalize="none"
+      :autofocus="autofocus"
```
Comment (samuveth) [inline src/components/BaseSearch.vue:59]: This doesn't work if I open the modal again after closing it. It should work every time like with our `BaseInput` `focus-on-mount` prop

File: src/components/SpaceProposalVotesModal.vue (hunk 5 above)
Comment (samuveth) [reply src/components/SpaceProposalVotesModal.vue:225]: Wdym?

File: src/components/SpaceProposalPage.vue (hunk 1 above)
Comment (samuveth) [reply src/components/SpaceProposalPage.vue:281]: >1

File: src/components/SpaceProposalVotesModal.vue (hunk 3 above)
Comment (samuveth) [reply src/components/SpaceProposalVotesModal.vue:145]: Yeah, lets see what designer Fab has to say about it.

File: src/components/SpaceProposalPage.vue (hunk 1 above)
Comment (samuveth) [reply src/components/SpaceProposalPage.vue:281]: Specially if it's larger amount of code

File: src/composables/useQuorum.ts [hunk 7]
```diff
@@ -9,7 +9,6 @@ interface QuorumProps {
   space: ExtendedSpace;
   proposal: Proposal;
   results: Results;
-  votes: Vote[];
```
Comment (samuveth) [inline src/composables/useQuorum.ts:12]: Why did you remove the votes prop is so mane components?

File: src/components/SpaceProposalVotesModal.vue [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        1
+      )[0]
+    );
+  } else {
+    uniqVotes.sort((a, b) => b.balance - a.balance);
+  }
+  return uniqVotes;
+});
+
+const titles = computed(() =>
```
Comment (samuveth) [inline src/components/SpaceProposalVotesModal.vue:50]: This should be inside the List component, instead of a prop

File: src/components/SpaceProposalVotesModal.vue [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+          v-for="(vote, i) in sortedVotes"
+          :key="i"
+          :vote="vote"
+          :profiles="profiles"
+          :space="space"
+          :proposal="proposal"
+          :titles="titles"
+          :class="{ '!border-0': i === 0 }"
+          :data-testid="`proposal-votes-list-item-${i}`"
+          @open-receipt-modal="openReceiptModal"
```
Comment (samuveth) [inline src/components/SpaceProposalVotesModal.vue:162]: Modal over modal is a bit ugly. We could use the `BasePopover` component to make it work like when you hover a user. We can use this solution for both instances of this component

File: src/composables/useQuorum.ts (hunk 7 above)
Comment (j0hnfl0w) [reply src/composables/useQuorum.ts:12]: We don't need them there, and we don't have votes in SpaceProposalPage no more, all logic is in SpaceProposalVotesList.

Comment (samuveth) [reply src/composables/useQuorum.ts:12]: Did you also check if any plugin is using `votes`?

Comment (j0hnfl0w) [reply src/composables/useQuorum.ts:12]: Didn't find any `votes` dependencies

Comment (samuveth) [reply src/composables/useQuorum.ts:12]: The votes prop is defined in multiple plugins. Can you add it back for plugins for now and create an issue to remove in another PR? 
image

File: src/components/SpaceProposalPage.vue [hunk 10]
```diff
@@ -205,14 +177,11 @@ onMounted(() => {
...
         <SpaceProposalVotesList
-          :loaded="loadedVotes"
           :space="space"
           :proposal="proposal"
-          :votes="votes"
           :strategies="strategies"
           :user-vote="userVote"
-          :loading-more="loadingMore"
-          @loadVotes="loadMore(loadMoreVotes)"
+          @votes="votes = $event"
```
Comment (samuveth) [inline src/components/SpaceProposalPage.vue:184]: Now when I refresh the page I can't see the votes block when votes after loading anymore. Probably better to use the composable instead of emitting

File: src/components/SpaceProposalVotesListItem.vue [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+        }"
+      >
+        {{
+          `${formatCompactNumber(vote.balance)} ${shorten(
+            proposal.symbol || space.symbol,
+            'symbol'
+          )}`
+        }}
+      </span>
+      <BasePopover>
```
Comment (samuveth) [inline src/components/SpaceProposalVotesListItem.vue:57]: We should use centered titles like everywhere else 
image

Comment (samuveth) [inline src/components/SpaceProposalVotesListItem.vue:57]: Should we use a centered title here? 
image

File: src/components/SpaceProposalVotesModal.vue [hunk 12]
```diff
@@ -0,0 +1,112 @@
+<script setup lang="ts">
```
Comment (samuveth) [inline src/components/SpaceProposalVotesModal.vue:1]: There should be a loading indicator when votes are loading after I open the modal

File: src/components/SpaceProposalVotesModal.vue [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    }
+  }
+);
+</script>
+
+<template>
+  <BaseModal
+    :open="open"
+    max-height="608px"
+    max-width="608px"
```
Comment (samuveth) [inline src/components/SpaceProposalVotesModal.vue:72]: To make it would with the our default width we can only show the avatar imagine without the address inside the modal.

File: src/components/SpaceProposalVotesModal.vue (hunk 3 above)
Comment (bonustrack) [reply src/components/SpaceProposalVotesModal.vue]: The modal should look like others modal, with same width. Also on small screen it should take the whole height, currently it's cut:
image

File: src/components/SpaceProposalVotesModal.vue (hunk 5 above)
Comment (samuveth) [reply src/components/SpaceProposalVotesModal.vue:104]: The infinite loading is super buggy for me https://www.loom.com/share/e3b88a02acb94edf94d50ec69240ae85

File: src/components/SpaceProposalPage.vue (hunk 10 above)
Comment (samuveth) [reply src/components/SpaceProposalPage.vue:184]: https://www.loom.com/share/4584d6c74bec45cab21e3f562ad256b9

File: src/components/SpaceProposalVotesModal.vue (hunk 12 above)
Comment (samuveth) [reply src/components/SpaceProposalVotesModal.vue:1]: https://www.loom.com/share/e4e0b2906e96458b9eebe942c8ceaf61

File: src/components/SpaceProposalVotesModal.vue (hunk 13 above)
Comment (samuveth) [reply src/components/SpaceProposalVotesModal.vue]: I would still recommend doing this as it all seems a bit tight when choices are longer. See screenshot:
image

File: src/components/SpaceProposalPage.vue (hunk 10 above)
Comment (j0hnfl0w) [reply src/components/SpaceProposalPage.vue:184]: Now we emit votes only to support prop for plugins

File: src/components/SpaceProposalVotesModal.vue (hunk 13 above)
Comment (j0hnfl0w) [reply src/components/SpaceProposalVotesModal.vue]: Still don't know why we need default width for all our modals, we have different content inside them...

File: src/components/SpaceProposalVotesModal.vue (hunk 5 above)
Comment (j0hnfl0w) [reply src/components/SpaceProposalVotesModal.vue:104]: Yep there was a throttle and small offset to bottom, fixed it.

File: src/components/SpaceProposalVotesModal.vue (hunk 13 above)
Comment (j0hnfl0w) [reply src/components/SpaceProposalVotesModal.vue]: image
it is so narrow that even the address barely fits in

**Comment by samuveth** [comment]
And https://www.loom.com/share/dffc89509ebc4eacadb1709d1699804d

File: src/components/BaseModal.vue [hunk 14]
```diff
@@ -35,7 +37,7 @@ onBeforeUnmount(() => {
           <slot name="header" />
         </div>
         <div class="modal-body">
-          <slot />
+          <slot :max-height="maxHeight" />
```
Comment (samuveth) [inline src/components/BaseModal.vue:40]: Do we still need this? And the maxWidth prop we can also remove I think

File: src/components/BaseSearch.vue [hunk 15]
```diff
@@ -57,7 +68,8 @@ function clearInput() {
       class="input w-full flex-auto border-none"
       @input="handleInput"
     />
-    <a @click="clearInput">
+    <slot name="after" :clearInput="clearInput" />
+    <a v-if="!$slots.after" @click="clearInput">
```
Comment (samuveth) [inline src/components/BaseSearch.vue:72]: We should use <i-ho-name /> like explained here https://www.notion.so/snapshotlabs/Getting-started-with-Snapshot-Ui-3ff7be7be0ac435797f198199d4f6e6e?pvs=4

BaseIcon is the old way that we still haven't fully refactored out of our code.

File: src/components/SpaceProposalVotesModal.vue [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+            :key="i"
+            :vote="vote"
+            :profiles="profiles"
+            :space="space"
+            :proposal="proposal"
+            :class="{ '!border-0': i === 0 }"
+            :data-testid="`proposal-votes-list-item-${i}`"
+            :hide-username="true"
+          />
+          <a
```
Comment (samuveth) [inline src/components/SpaceProposalVotesModal.vue:140]: Why do we need `<a>` around the loading spinner?

File: src/components/SpaceProposalVotesModal.vue [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+          style="width: 80%; height: 20px"
+        />
+        <div class="lazy-loading rounded-md" style="width: 50%; height: 20px" />
+      </div>
+      <div
+        v-if="loadedVotes && votesNotFound"
+        class="flex flex-row content-center items-center justify-center py-6"
+        :style="{ height: maxHeight }"
+      >
+        <span>{{ $t('noResultsFound') }}</span>
```
Comment (samuveth) [inline src/components/SpaceProposalVotesModal.vue:120]: This should be at the top instead of vertically centered. Example
image

File: src/components/SpaceProposalVotesModal.vue [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+        >
+          <template
+            v-if="isENS && (resolvingENS || wrongENS)"
+            #after="{ clearInput }"
+          >
+            <LoadingSpinner
+              v-if="isENS && resolvingENS && !wrongENS"
+              class="pb-[3px]"
+            />
+            <i-ho-x
```
Comment (samuveth) [inline src/components/SpaceProposalVotesModal.vue:94]: This doesn't seem so obvious, I think it would be better to always show `no results found` when something was entered into the search input but no vote or ens address was found.

File: src/components/SpaceProposalVotesListItem.vue [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+<template>
+  <div class="flex items-center border-t px-3 py-[14px]">
+    <BaseUser
+      :key="vote.voter"
+      :profile="profiles[vote.voter]"
+      :address="vote.voter"
+      :space="space"
+      :proposal="proposal"
+      :hide-username="hideUsername"
+      width-class="w-[110px] min-w-[110px] xs:w-[130px] xs:min-w-[130px] text-left"
```
Comment (samuveth) [inline src/components/SpaceProposalVotesListItem.vue:41]:
```suggestion
       :width-class="
        hideUsername
          ? 'w-[50px]'
          : 'w-[110px] min-w-[110px] xs:w-[130px] xs:min-w-[130px] text-left'
      "
```
image

File: src/components/BaseModal.vue (hunk 14 above)
Comment (j0hnfl0w) [reply src/components/BaseModal.vue:40]: We still need maxHeight slot prop for infinite scroll container, for explicit height. Removed maxWidth prop.

File: src/components/SpaceProposalVotesListItem.vue [hunk 20]
```diff
@@ -30,19 +24,14 @@ const { formatCompactNumber } = useIntl();
...
       :address="vote.voter"
       :space="space"
       :proposal="proposal"
-      :hide-username="hideUsername"
-      :width-class="
-        hideUsername
-          ? 'w-[50px]'
-          : 'w-[110px] min-w-[110px] xs:w-[130px] xs:min-w-[130px] text-left'
-      "
+      :width-class="'w-[110px] min-w-[110px] xs:w-[130px] xs:min-w-[130px] text-left '"
```
Comment (j0hnfl0w) [inline src/components/SpaceProposalVotesListItem.vue:34]: why this class is still dynamic?

Comment (samuveth) [reply src/components/SpaceProposalVotesListItem.vue:34]: Ah, yes will remove
